internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,301 @@
+	sl.mlJoinCh = make(chan struct{}, 5000*3)
+
+	// ChannelEventDelegate hint that it should not block, so make mlEventCh 'big'
+	sl.mlEventCh = make(chan memberlist.NodeEvent, 1024)
+	mconfig.Events = &memberlist.ChannelEventDelegate{Ch: sl.mlEventCh}
+	var err error
+	sl.mList, err = memberlist.Create(mconfig)
+	if err != nil {
+		sl.l.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) error {
+	sl.client = client
+
+	if sl.mList == nil {
+		return nil
+	}
+
+	go sl.memberlistWatchEvents()
+	go sl.mlJoinLoop()
+
+	return nil
+}
+
+// updateMlIPs runs forever updating the sl.mlSpeakersIPs slice with the
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
+	for {
+		// Create the ticker at the start of each loop, so we wait at
+		// least the ticker interval before the ticker case is true.
+		// If we create it outside of the loop, it will tick every 5
+		// minutes and if the blocking call happens to take longer, the
+		// next tick will be really soon and we may overload the API
+		// server.
+		ticker := time.NewTicker(5 * time.Minute)
+
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			// This blocks until the apiserver responds.
+			iplist, err := sl.mlSpeakers()
+			if err != nil {
+				continue
+			}
+
+			sl.mlMux.Lock()
+			sl.mlSpeakersIPs = iplist
+			sl.mlMux.Unlock()
+		}
+	}
+}
+
+func (sl *SpeakerList) mlSpeakers() ([]string, error) {
+	// This call blocks until we get a response of the API server.
+	// In the client-go version we are using, there is no way to use a
+	// context or timeout.
+	// TODO: When updating client-go, this code can be simplified by using a
+	// context in the following way:
+	// If we use a context, we can specify a timeout and there is no need to
+	// have the speakersIPsLoop() running in the background. This can just
+	// be called from mlJoin() with a context (timeout) and if doesn't
+	// answer, use sl.mlSpeakersIPs.  If it doesn't time out, just update
+	// sl.mlSpeakersIPs with the answer and continue to do the join.
+	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
+	if err != nil {
+		sl.l.Log("op", "Member Detection", "error", err, "msg", "failed to get PodsIPs")
+		return nil, err
+	}
+
+	return iplist, nil
+}
+
+func (sl *SpeakerList) mlJoinLoop() {
+	iplist, err := sl.mlSpeakers()
+	if err != nil {
+		sl.l.Log("op", "startup", "error", err, "msg", "failed to get PodsIPs")
+		iplist = nil
+	}
+
+	// Initialize sl.mlSpeakersIPs.
+	sl.mlMux.Lock()
+	sl.mlSpeakersIPs = iplist
+	sl.mlMux.Unlock()
+
+	// Update mlSpeakersIPs in the background.
+	go sl.speakersIPsLoop()
+
+	// Every one minute, try to rejoin.
```

Comment:
Why loop based on a timer here? I suggest we look into the following alternative (if possible):

- Only try to rejoin when there are "unjoined" members.
- Attempts a rejoin ASAP following a failure event, possibly backing off exponentially.

Reply:
Makes sense. We can make a node leave event trigger a re-join. But I don't think the second point helps.

For example, regarding the second point, N nodes leave, we try to join them all. Now only N-X are not joined and we do that until when? An exponential backoff? Then, after some time without those nodes joining, another node leaves, and you need to reset the backoff, right? It starts to feel complicated. Instead of exponential backoff, just retry every minute if there is something to do.

If you do it once a minute, only when there are nodes to join (i.e. it checks members and only tries to join when some IP is not member). 

I agree with doing things the way you propose, I just didn't find any easy/simple way to do that.

Reply:
@johananl what you want is more or less in #595, so let's clean this PR, merge it and have the patch release, then we can review 1 patch at a time #595

Reply:
@johananl does what I said makes sense for you at least for now?

When we have a watch mechanism, I think, the only logic here will be to join the IPs that are not part of the cluster but returned by the watch mechanism. Maybe every one minute, maybe with some exponential backoff, we will need to see how the watch mechanism looks like.

Does it makes sense to leave it as it is for now? Or what would you change (without a watch)?